Проект 8 [Дегтерев, Орешников]

doreshnikov opened the pull request.

Commit:
test commit

doreshnikov committed:
script added

elmanelman commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+    email       TEXT NOT NULL UNIQUE,
+    phone       TEXT NOT NULL UNIQUE,
+    sex         TEXT,
+    birthday    DATE,
```
Возможно, стоило бы добавить `CHECK` на адекватность даты рождения.

elmanelman commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+    second_name TEXT NOT NULL,
+    email       TEXT NOT NULL UNIQUE,
+    phone       TEXT NOT NULL UNIQUE,
+    sex         TEXT,
```
Хорошо, если согласованность полов обеспечивается логикой приложения, которое работает с базой. Но мне кажется, что стоит добавить `CHECK` или даже вынести список полов в отдельный тип (возможно даже в отдельную таблицу, если следовать трендам).

OrangeFlag replied:
Отдельная таблица или check

elmanelman commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+-- Список жанров
+CREATE TYPE GENRE AS ENUM (
+    'фестиваль'
+    );
```
Думаю, лучше было бы использовать таблицу.

elmanelman commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+    id         INT PRIMARY KEY,
+    name       TEXT NOT NULL,
+    address    TEXT NOT NULL,
+    gps        TEXT,
```
Возможно, стоит использовать тип, хоть как-то обеспечивающий корректность (например, `point`).

dyupon replied:
not null? Даже в примерах приведены координаты

OrangeFlag replied:
gps лучше хранить в point или подключать PostGIS, но для этой задачи последнее слишком конечно

elmanelman commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+    -- внешний ключ на таблицу Countries, связь many-to-one (N:1)
+    country_id     INT  NOT NULL references Countries (id),
+    address        TEXT NOT NULL,
+    gps            TEXT,
```
Возможно, стоит использовать тип, хоть как-то обеспечивающий корректность (например, `point`).

elmanelman commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+    -- параметры количества комнат\кроватей\соседей должны быть положительными
+    room_count     INT CHECK (room_count > 0),
+    bed_count      INT CHECK (bed_count > 0),
+    max_roommates  INT CHECK (max_roommates > 0),
```
Наверное, здесь `NOT NULL`?

dyupon replied:
> пользователь **должен** указать

elmanelman commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+    -- внешний ключ на BookingReviews, отношение many-to-many (N:M)
+    review_id INT references BookingReviews (application_id),
+    param     BookingReviewParams,
+    score     INT CHECK (score > 0 and score <= 5),
```
Наверное, здесь `NOT NULL`?

doreshnikov replied:
Выше да, а здесь подразумевается, что оценка выставляется не сразу, поэтому до какого-то момента она `NULL`

elmanelman commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+(
+    -- внешний ключ на Application, отношение one-to-one (не более одного отзыва на заявку)
+    application_id INT PRIMARY KEY references Applications (id),
+    review         TEXT,
+    -- оценка может принимать значение от 1 до 5
+    score          INT CHECK (score > 0 and score <= 5)
```
Наверное, здесь `NOT NULL`?

doreshnikov replied:
Аналогично прошлому комментарию, оценка выставляется только по прошествию какого-то времени

elmanelman commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+    -- даты должны быть корректно упорядочены
+    end_date        DATE    NOT NULL CHECK (end_date > start_date),
+    -- соседей неотрицательное число
+    roommates_count INT     NOT NULL CHECK (roommates_count >= 0),
```
Видимо, в задании нужно было указать фактическое число проживающих, так как пользователь, оформляющий заявку, может сам не проживать (а может по другим причинам). Отсюда следует, что в `CHECK` лучше проверять строгое неравенство.

elmanelman commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+    user_id         INT references Users (id),
+    start_date      DATE    NOT NULL,
+    -- даты должны быть корректно упорядочены
+    end_date        DATE    NOT NULL CHECK (end_date > start_date),
```
Предположу, что могут быть заявки на краткосрочное проживание (меньше суток), тогда лучше проверять нестрогое неравенство (или же использовать `timestamp`, чтобы уточнить время).

elmanelman commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+-- PRIMARY KEY отвечает за отсутствие одинаковых записей в таблице
+CREATE TABLE Users_Apartments
+(
+    -- merge-table, реализующая отношение many-to-many между владельцами и сдающимися апартаментами
```
А тут точно many-to-many? Если да, то прошу указать на место в задании. Мне всё-таки кажется, что за конкретным жильем лучше закреплять одного владельца.

doreshnikov replied:
Да, логично, но мы подумали, что нет смысла запрещать нескольких владельцев одних апартаментов. Хотя, конечно, согласен, что это накладывает некоторые сложности

doreshnikov replied:
@dbarashev эта проблема отчасти возникла на предыдущей итерации и была не исправлена на последней. В предыдущей итерации была заведена merge-таблица `Renter` с референсами на `Apartments` и `Users`, но к такому архитектурному решению не было комментариев.

Поэтому мы логично подумали, что раз есть отдельная таблица с референсами на две данные, то это many-to-many. Я сейчас закоммитил изменения, поменял ключи, но саму таблицу оставил.

elmanelman commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+-- Требуем NOT NULL для минимальной значимой информации
+CREATE TABLE Event
+(
+    id         INT PRIMARY KEY,
```
Дело вкуса, но я бы использовал какой-нибудь `serial` тип, чтобы не заботиться о возможных проблемах с `id`.

elmanelman reviewed (commented):
С точки зрения требований работа отличная, в комментариях в основном предлагаю небольшие исправления.

alefly commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+-- Email и телефон уникальные, чтобы не разрешать мультиаккаунт для одного и того же человека
+CREATE TABLE Users
+(
+    id          INT PRIMARY KEY,
```
Здесь и в других таблицах можно сделать id положительным и автоинкрементным, изменив int на serial или использовать auto_increment. Если не ошибаюсь, без этого будет очень неудобно вставлять данные)

P.S. Я отключила комментарии других интервьюеров, поэтому сори за повторения, если они будут

dyupon replied:
А откуда здесь берется id? Вероятно, оно должно быть SERIAL или что-то типа того? Кажется, ко всем таблицам ниже такой же вопрос

OrangeFlag replied:
SERIAL

P.S. Буду краток, т.к. кажется это все уже расписывали, что-то новое уже буду писать развернуто

alefly commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+    -- внешний ключ на таблицу Countries, связь many-to-one (N:1)
+    country_id     INT  NOT NULL references Countries (id),
+    address        TEXT NOT NULL,
+    gps            TEXT,
```
В разных системах координаты могут обозначаться по-разному, например целая часть от дробной может отделяться запятой или точкой. Не знаю как решить эту проблему кроме как вынести gps в три вещественных атрибута - долгота, широта и высота

dbarashev replied:
высота не требуется

alefly commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+
+-- Таблица (merge-table) владельцев жилья - пользователь с этим id (user_id) владеет жильём с id (appartment_id)
+-- PRIMARY KEY отвечает за отсутствие одинаковых записей в таблице
+CREATE TABLE Users_Apartments
```
Одни и те же апартаменты наверное всё-таки не могут сдавать два разных человека. Иначе, если это всё-таки возможно, то в заявке на бронирование стоит указывать какому из арендодателей платить денежки, или в каких долях делить их между арендодателями. Без этого получается интересная схема развода на деньги ;)

alefly commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+    -- внешний ключ на BookingReviews, отношение many-to-many (N:M)
+    review_id INT references BookingReviews (application_id),
+    param     BookingReviewParams,
+    score     INT CHECK (score > 0 and score <= 5),
```
Здесь точно not null, иначе получается строка таблицы с параметром, но без оценки, что не несет никакой информативности. Если пользователь не поставил оценку по некоторым параметрам, то лучше наверное вообще не сохранять информацию об этом, чтобы не замедлять сбор статистики оценок

dyupon replied:
not null

OrangeFlag replied:
not null

p.s. честно не уверен, что стоит писать тоже, что и 3 других ревьювера(

dyupon commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+    id   INT PRIMARY KEY,
+    name TEXT NOT NULL UNIQUE,
+    -- ограничение на неотрицательный сервисный сбор
+    tax  INT  NOT NULL CHECK (tax >= 0)
```
Комиссия вполне может составлять полтора таллера, может, на надо делать поле интом? Как и в других полях с ценами в схеме. JFYI: в постгресе есть даже тип данных а ля currency (кажется, основная его фишка в отличие от какого-нибудь float в том, что можно задавать точность, с которой будут отображаться ценники в базе)

dyupon commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+    -- внешний ключ на Apartments, отношение many-to-one (M:1)
+    apartment_id INT references Apartments (id),
+    -- количество недель в году можно ограничить числом 53
+    week         INT NOT NULL CHECK (week > 0 and week <= 53),
```
здесь, конечно, может быть проблема при заключении договора аренды на длительный срок от года

dyupon commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+-- PRIMARY KEY отвечает за отсутствие одинаковых записей в таблице
+CREATE TABLE Users_Apartments
+(
+    -- merge-table, реализующая отношение many-to-many между владельцами и сдающимися апартаментами
```
Не похоже, что многие ко многим -- иначе получалось бы, что оценки жильцам надо ставить по отдельности нескольким хозяевам за одно и то же жилье

dyupon commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+-- PRIMARY KEY состоит из двух полей, потому что мы не хотим допустить две разные оценки по одному параметру в одном review, но у одного review могут быть оценки по разным параметрам и наоборот
+CREATE TABLE BookingReviews_BookingReviewParams
+(
+    -- внешний ключ на BookingReviews, отношение many-to-many (N:M)
```
N:1? Ревью одно, оценок по разным критериям в одном ревью несколько

dyupon commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+
+-- Список жанров
+CREATE TYPE GENRE AS ENUM (
+    'фестиваль'
```
еще там были "спорт" и "пляж"

dyupon commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+(
+    -- внешний ключ на Application, отношение one-to-one (не более одного отзыва на заявку)
+    application_id INT PRIMARY KEY references Applications (id),
+    review         TEXT,
+    -- оценка может принимать значение от 1 до 5
+    score          INT CHECK (score > 0 and score <= 5)
```
что-то из этого явно не null

dyupon reviewed (commented):
Есть несколько весомых неточностей, но в целом нормально

OrangeFlag commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+    id          INT PRIMARY KEY,
+    first_name  TEXT NOT NULL,
+    second_name TEXT NOT NULL,
+    email       TEXT NOT NULL UNIQUE,
+    phone       TEXT NOT NULL UNIQUE,
```
CHECK (email ~'^[^@ \t\r\n]+@[^@ \t\r\n]+\.[^@ \t\r\n]+$')
CHECK (phone ~ '^\+[1-9]\d{1,14}$' )

dbarashev replied:
Всё [не так уж и просто](https://stackoverflow.com/questions/201323/how-to-validate-an-email-address-using-a-regular-expression) с епочтой.

OrangeFlag commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+    phone       TEXT NOT NULL UNIQUE,
+    sex         TEXT,
+    birthday    DATE,
+    photo       TEXT
```
Возможно лучше photo_url?

OrangeFlag commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+(
+    id              INT PRIMARY KEY,
+    -- внешний ключ на апартаменты, отношение many-to-one (N:1) (бывает много заявок на одни апартаменты)
+    apartment_id    INT references Apartments (id),
+    -- внешний ключ на Users, отношение many-to-one (N:1) (аналогично)
+    user_id         INT references Users (id),
```
NOT NULL

OrangeFlag commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+
+-- Список параметров оценки жилья
+-- ENUM со списком всех параметров для оценки
+CREATE TYPE BookingReviewParams AS ENUM (
```
Может быть лучше таблицей?

OrangeFlag commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+(
+    -- внешний ключ на BookingReviews, отношение many-to-many (N:M)
+    review_id INT references BookingReviews (application_id),
+    param     BookingReviewParams,
```
И если BookingReviewParams таблица, то тут references

OrangeFlag commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+);
+
+-- Список жанров
+CREATE TYPE GENRE AS ENUM (
```
Таблица была бы лучше, да

OrangeFlag commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+    gps        TEXT,
+    date_start DATE NOT NULL,
+    -- событие либо длится неопределенное время, либо неотрицательное
+    date_end   DATE CHECK (date_end >= date_start),
```
Как писали выше лучше задать продолжительность

OrangeFlag commented on project8/project8.sql:
```diff
@@ -0,0 +1,168 @@
+    name       TEXT NOT NULL,
+    address    TEXT NOT NULL,
+    gps        TEXT,
+    date_start DATE NOT NULL,
```
Возможно DATETIME
`в какой-то момент или интервал времени`

OrangeFlag reviewed (commented):
Хорошая работа!

dbarashev commented:
@degterevz @doreshnikov отметил эмоджей с глазками комментарии к недостаткам, причину и итерацию появления которых нужно выяснить

doreshnikov committed:
Fixed all issues:
- changed INT to SERIAL
- changed some ENUMs to TABLEs
- added NOT NULLs to required fields
- added some TEXT CHECKs and selected better types

doreshnikov committed:
small comment fix

doreshnikov reviewed (commented):
@dbarashev ответил на отмеченный "глазками" пункт, закоммитил исправления по всем (вроде как) замечаниям

Commit:
some code

doreshnikov committed:
Merge remote-tracking branch 'origin/project0' into project8

doreshnikov committed:
removed merge-table for N:1 relation

doreshnikov committed:
added run() to webapp's __main__, added 'cd' to project0 in main.sh

doreshnikov committed:
Merge branch 'project8' of https://github.com/dbms-class/csc-2020-team-all into project8

doreshnikov committed:
Added all endpoints:
- countries
- apartments
- update_price

doreshnikov committed:
moved webapp and conn to project8 folder

doreshnikov committed:
merge conflict resolved? on project8

doreshnikov committed:
removed project0 from branch project8

doreshnikov committed:
added static.py back

doreshnikov committed:
added report

doreshnikov committed:
slightly modified static.py

doreshnikov committed:
added new  method for apartments search

Commit:
Merge remote-tracking branch 'origin/project8' into project8

doreshnikov committed:
Merge performed

doreshnikov committed:
added report for previous and current week

doreshnikov committed:
fixed min_price and max_price handling

doreshnikov committed:
minor cosmetic fix

doreshnikov committed:
minor cosmetic report fix

doreshnikov committed:
Merge branch 'project8' of https://github.com/dbms-class/csc-2020-team-all into project8

dbarashev commented on project8/project8.sql:
```diff
@@ -0,0 +1,175 @@
+    -- внешний ключ на таблицу Users, связь many-to-one (N:1)
+    owner_id       INT  NOT NULL references Users (id),
+    address        TEXT NOT NULL,
+    gps            POINT,
```
Точка на плоскости таки не то же самое, что точка на сфере. Арифметика другая, ограничения другие.

dbarashev commented on project8/project8.sql:
```diff
@@ -0,0 +1,175 @@
+    apartment_id    INT       NOT NULL references Apartments (id),
+    -- внешний ключ на Users, отношение many-to-one (N:1) (аналогично)
+    user_id         INT       NOT NULL references Users (id),
+    start_date      TIMESTAMP NOT NULL,
```
В ТЗ везде "даты проживания", а не время с точностью до микросекунд.

Commit:
done

lissrbay committed:
Fix get_price method, write part of appt_sale, add test data

dbarashev committed:
работа 2020-12-03

lissrbay committed:
Merge two branch project8

ioart committed:
2020-12-10

dbarashev committed:
2020-12-10 report
